In [1]:
import pandas as pd

#using a relative path
file_path = "../data/AmesHousing.csv"
df = pd.read_csv(file_path)

df.head()

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot Shape Land Contour  ... Pool Area Pool QC  Fence Misc Feature  \
0   NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
1   NaN       Reg          Lvl  ...         0     NaN  MnPrv          NaN   
2   NaN       IR1          Lvl  ...         0     NaN    NaN         Gar2   
3   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
4   NaN       IR1          Lvl  ...         0     NaN  MnPrv          NaN   

  Misc Val Mo Sold Yr Sold Sale Type  Sale Condition  SalePrice  
0        0       5    2010       WD           Normal     215000  
1        0       6    2010       WD           Normal     105000  
2    12500       6    2010       WD           Normal     172000  
3        0       4    2010       WD           Normal     244000  
4        0       3    2010       WD           Normal     189900  

[5 rows x 82 columns]

In [2]:
#view the type of columns and datatypes
df.info()

# Quick descriptive statistics for numeric columns
df.describe().T


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order            2930 non-null   int64  
 1   PID              2930 non-null   int64  
 2   MS SubClass      2930 non-null   int64  
 3   MS Zoning        2930 non-null   object 
 4   Lot Frontage     2440 non-null   float64
 5   Lot Area         2930 non-null   int64  
 6   Street           2930 non-null   object 
 7   Alley            198 non-null    object 
 8   Lot Shape        2930 non-null   object 
 9   Land Contour     2930 non-null   object 
 10  Utilities        2930 non-null   object 
 11  Lot Config       2930 non-null   object 
 12  Land Slope       2930 non-null   object 
 13  Neighborhood     2930 non-null   object 
 14  Condition 1      2930 non-null   object 
 15  Condition 2      2930 non-null   object 
 16  Bldg Type        2930 non-null   object 
 17  House Style   

count          mean           std          min  \
Order            2930.0  1.465500e+03  8.459625e+02          1.0   
PID              2930.0  7.144645e+08  1.887308e+08  526301100.0   
MS SubClass      2930.0  5.738737e+01  4.263802e+01         20.0   
Lot Frontage     2440.0  6.922459e+01  2.336533e+01         21.0   
Lot Area         2930.0  1.014792e+04  7.880018e+03       1300.0   
Overall Qual     2930.0  6.094881e+00  1.411026e+00          1.0   
Overall Cond     2930.0  5.563140e+00  1.111537e+00          1.0   
Year Built       2930.0  1.971356e+03  3.024536e+01       1872.0   
Year Remod/Add   2930.0  1.984267e+03  2.086029e+01       1950.0   
Mas Vnr Area     2907.0  1.018968e+02  1.791126e+02          0.0   
BsmtFin SF 1     2929.0  4.426296e+02  4.555908e+02          0.0   
BsmtFin SF 2     2929.0  4.972243e+01  1.691685e+02          0.0   
Bsmt Unf SF      2929.0  5.592625e+02  4.394942e+02          0.0   
Total Bsmt SF    2929.0  1.051615e+03  4.406151e+02          0.0   
1st Flr SF       2930.0  1.159558e+03  3.918909e+02        334.0   
2nd Flr SF       2930.0  3.354560e+02  4.283957e+02          0.0   
Low Qual Fin SF  2930.0  4.676792e+00  4.631051e+01          0.0   
Gr Liv Area      2930.0  1.499690e+03  5.055089e+02        334.0   
Bsmt Full Bath   2928.0  4.313525e-01  5.248202e-01          0.0   
Bsmt Half Bath   2928.0  6.113388e-02  2.452536e-01          0.0   
Full Bath        2930.0  1.566553e+00  5.529406e-01          0.0   
Half Bath        2930.0  3.795222e-01  5.026293e-01          0.0   
Bedroom AbvGr    2930.0  2.854266e+00  8.277311e-01          0.0   
Kitchen AbvGr    2930.0  1.044369e+00  2.140762e-01          0.0   
TotRms AbvGrd    2930.0  6.443003e+00  1.572964e+00          2.0   
Fireplaces       2930.0  5.993174e-01  6.479209e-01          0.0   
Garage Yr Blt    2771.0  1.978132e+03  2.552841e+01       1895.0   
Garage Cars      2929.0  1.766815e+00  7.605664e-01          0.0   
Garage Area      2929.0  4.728197e+02  2.150465e+02          0.0   
Wood Deck SF     2930.0  9.375188e+01  1.263616e+02          0.0   
Open Porch SF    2930.0  4.753345e+01  6.748340e+01          0.0   
Enclosed Porch   2930.0  2.301160e+01  6.413906e+01          0.0   
3Ssn Porch       2930.0  2.592491e+00  2.514133e+01          0.0   
Screen Porch     2930.0  1.600205e+01  5.608737e+01          0.0   
Pool Area        2930.0  2.243345e+00  3.559718e+01          0.0   
Misc Val         2930.0  5.063515e+01  5.663443e+02          0.0   
Mo Sold          2930.0  6.216041e+00  2.714492e+00          1.0   
Yr Sold          2930.0  2.007790e+03  1.316613e+00       2006.0   
SalePrice        2930.0  1.807961e+05  7.988669e+04      12789.0   

                          25%          50%           75%           max  
Order            7.332500e+02       1465.5  2.197750e+03  2.930000e+03  
PID              5.284770e+08  535453620.0  9.071811e+08  1.007100e+09  
MS SubClass      2.000000e+01         50.0  7.000000e+01  1.900000e+02  
Lot Frontage     5.800000e+01         68.0  8.000000e+01  3.130000e+02  
Lot Area         7.440250e+03       9436.5  1.155525e+04  2.152450e+05  
Overall Qual     5.000000e+00          6.0  7.000000e+00  1.000000e+01  
Overall Cond     5.000000e+00          5.0  6.000000e+00  9.000000e+00  
Year Built       1.954000e+03       1973.0  2.001000e+03  2.010000e+03  
Year Remod/Add   1.965000e+03       1993.0  2.004000e+03  2.010000e+03  
Mas Vnr Area     0.000000e+00          0.0  1.640000e+02  1.600000e+03  
BsmtFin SF 1     0.000000e+00        370.0  7.340000e+02  5.644000e+03  
BsmtFin SF 2     0.000000e+00          0.0  0.000000e+00  1.526000e+03  
Bsmt Unf SF      2.190000e+02        466.0  8.020000e+02  2.336000e+03  
Total Bsmt SF    7.930000e+02        990.0  1.302000e+03  6.110000e+03  
1st Flr SF       8.762500e+02       1084.0  1.384000e+03  5.095000e+03  
2nd Flr SF       0.000000e+00          0.0  7.037500e+02  2.065000e+03  
Low Qual Fin SF  0.000000e+00          0.0  0.000000e+00

In [3]:
# For non-numeric columns
cat_cols = df.select_dtypes(include=['object', 'category']).columns
print("Number of categorical columns:", len(cat_cols))
print(cat_cols.tolist())

for col in cat_cols:
    print(f"--- {col} ---")
    print(df[col].value_counts(dropna=False).head(10))  
    print("\n")


Number of categorical columns: 43
['MS Zoning', 'Street', 'Alley', 'Lot Shape', 'Land Contour', 'Utilities', 'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1', 'Condition 2', 'Bldg Type', 'House Style', 'Roof Style', 'Roof Matl', 'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type', 'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin Type 2', 'Heating', 'Heating QC', 'Central Air', 'Electrical', 'Kitchen Qual', 'Functional', 'Fireplace Qu', 'Garage Type', 'Garage Finish', 'Garage Qual', 'Garage Cond', 'Paved Drive', 'Pool QC', 'Fence', 'Misc Feature', 'Sale Type', 'Sale Condition']
--- MS Zoning ---
MS Zoning
RL         2273
RM          462
FV          139
RH           27
C (all)      25
I (all)       2
A (agr)       2
Name: count, dtype: int64


--- Street ---
Street
Pave    2918
Grvl      12
Name: count, dtype: int64


--- Alley ---
Alley
NaN     2732
Grvl     120
Pave      78
Name: count, dtype: int64


--- Lot Shape ---


The most common observations of categorical variables include RL for MS Zoning, Pave for street, and NaN meaning null for Ally, thus meaning that Grvl is the only option for Alley and most observations do not have one.

In [4]:
# outlier function
def find_outliers_iqr(series):
    Q1 = series.quantile(0.25)
    Q3 = series.quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    return series[(series < lower) | (series > upper)]

#checking salesprice
outliers = find_outliers_iqr(df['SalePrice'])
print(f"Outliers in SalePrice: {len(outliers)}")

# Remove outliers in SalePrice
Q1 = df['SalePrice'].quantile(0.25)
Q3 = df['SalePrice'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

df = df[(df['SalePrice'] >= lower) & (df['SalePrice'] <= upper)]

print("Shape after outlier removal:", df.shape)

Outliers in SalePrice: 137
Shape after outlier removal: (2793, 82)


### Exploration Cell Provides the Following:
##### total missing cells
##### any columns that are completely empty
##### percentage of missing data overall
##### columns with more than 50% of values missing
##### quick check for numeric columns missing
##### proof that no 'critical' columns (e.g., SalePrice) are missing

In [8]:
print("Total missing values:", df.isna().sum().sum())
print("Columns entirely null:", df.columns[df.isna().all()].tolist())
print("Overall missing %:", round(df.isna().mean().mean() * 100, 2))

#high missing was agreed upon term of 50% missing data points in variable
high_missing = df.isna().mean() * 100
print("Columns >50% missing:\n", high_missing[high_missing > 50])

numeric_missing = df.select_dtypes(include="number").isna().sum().sum()
print("Numeric missing count:", numeric_missing)

critical_cols = ['SalePrice', 'Gr Liv Area', 'Overall Qual']
print(df[critical_cols].isna().sum())

Total missing values: 15180
Columns entirely null: []
Overall missing %: 6.63
Columns >50% missing:
 Alley           92.946652
Mas Vnr Type    62.799857
Fireplace Qu    50.841389
Pool QC         99.606158
Fence           79.735052
Misc Feature    96.204798
dtype: float64
Numeric missing count: 675
SalePrice       0
Gr Liv Area     0
Overall Qual    0
dtype: int64


Correlation Modeling Section

In [6]:

#only pull numeric columns
numeric_df = df.select_dtypes(include=['number'])

# Compute Pearson correlation matrix
corr_matrix = numeric_df.corr(method='pearson')

# Check shape (should be around 30–40 columns)
print("Number of numeric columns:", len(numeric_df.columns))
corr_matrix.head()

Number of numeric columns: 39


Order       PID  MS SubClass  Lot Frontage  Lot Area  \
Order         1.000000  0.171452     0.008400     -0.000262  0.045978   
PID           0.171452  1.000000    -0.013018     -0.059906  0.060197   
MS SubClass   0.008400 -0.013018     1.000000     -0.421781 -0.221528   
Lot Frontage -0.000262 -0.059906    -0.421781      1.000000  0.634971   
Lot Area      0.045978  0.060197    -0.221528      0.634971  1.000000   

              Overall Qual  Overall Cond  Year Built  Year Remod/Add  \
Order            -0.048280     -0.016734   -0.047558       -0.074479   
PID              -0.221568      0.088129   -0.321410       -0.130652   
MS SubClass       0.078277     -0.078001    0.055835        0.060076   
Lot Frontage      0.132416     -0.051138    0.075405        0.049520   
Lot Area          0.049086     -0.028048    0.001219        0.006645   

              Mas Vnr Area  ...  Wood Deck SF  Open Porch SF  Enclosed Porch  \
Order            -0.025423  ...     -0.019996       0.022077        0.027384   
PID              -0.189574  ...     -0.023289      -0.051773        0.160196   
MS SubClass       0.037141  ...     -0.004294      -0.004982       -0.027217   
Lot Frontage      0.149449  ...      0.088550       0.137277        0.025335   
Lot Area          0.100005  ...      0.153814       0.110659        0.039855   

              3Ssn Porch  Screen Porch  Pool Area  Misc Val   Mo Sold  \
Order          -0.024772      0.001389   0.054742 -0.006492  0.133031   
PID            -0.027784     -0.026975   0.001998 -0.011566 -0.052433   
MS SubClass    -0.040552     -0.055330  -0.003576 -0.030942  0.002339   
Lot Frontage    0.030959      0.081790   0.167260  0.051946  0.010638   
Lot Area        0.024234      0.045603   0.113701  0.085852  0.002607   

               Yr Sold  SalePrice  
Order        -0.976941  -0.029564  
PID           0.006985  -0.200899  
MS SubClass  -0.018039  -0.053960  
Lot Frontage -0.002767   0.300798  
Lot Area     -0.031716   0.241337  

[5 rows x 39 columns]

Only the top 10 Correlated Variables

In [7]:
corr_pairs = corr_matrix.unstack().reset_index()
corr_pairs.columns = ['Variable_1', 'Variable_2', 'Correlation']

# Drop self-correlations (a variable with itself)
corr_pairs = corr_pairs[corr_pairs['Variable_1'] != corr_pairs['Variable_2']]

# Keep only one of each pair (e.g., A-B but not B-A)
corr_pairs['Sorted'] = corr_pairs.apply(lambda x: tuple(sorted([x['Variable_1'], x['Variable_2']])), axis=1)
corr_pairs = corr_pairs.drop_duplicates(subset='Sorted').drop(columns='Sorted')

# Sort by absolute correlation
corr_pairs['AbsCorrelation'] = corr_pairs['Correlation'].abs()
top_corr = corr_pairs.sort_values(by='AbsCorrelation', ascending=False).head(10)

top_corr


Variable_1     Variable_2  Correlation  AbsCorrelation
37            Order        Yr Sold    -0.976941        0.976941
1081    Garage Cars    Garage Area     0.884185        0.884185
299      Year Built  Garage Yr Blt     0.828100        0.828100
687     Gr Liv Area  TotRms AbvGrd     0.796303        0.796303
233    Overall Qual      SalePrice     0.785878        0.785878
521   Total Bsmt SF     1st Flr SF     0.756622        0.756622
882   Bedroom AbvGr  TotRms AbvGrd     0.693327        0.693327
602      2nd Flr SF    Gr Liv Area     0.661542        0.661542
701     Gr Liv Area      SalePrice     0.652255        0.652255
121    Lot Frontage       Lot Area     0.634971        0.634971

We the strongest correlation is between Order and Yr Sold with a -97% R. Meaning a very strong negative linear relationship.

The dataset has 2,930 rows and 82 variables. Roughly 6.55% of the data is missing, mostly from Pool QC (99.6%), Misc Feature (96.4%), Alley (93.2%), and Fence (80.5%). Those fields rarely apply to most homes. After removing 137 outliers, 2,793 clean records remain.

Overall Quality, Gr Liv Area, and Garage Cars have no missing values. These three features strongly influence SalePrice, with correlations of 0.80, 0.71, and roughly 0.64, respectively. Year Built and Garage Yr Blt correlate at 0.83, showing that newer homes tend to have newer garages.

The dataset is clean and ready for modeling. Larger, higher-quality, and newer houses generally sell for more.